### PR TITLE
Commander Core: Cleanup and add read error checking

### PR DESCRIPTION
**Pull request by ParkerMc**
Clean up Commander Core code and documentation.

Ensure we received the right data on read.  
The first two bytes of data in response to the read command appear to be a data type so now those are being checked.

Related: #218 

---

Checklist:

- [x] Adhere to the [development process]
- [x] Conform to the [style guide]
- [x] Verify that the changes work as expected on real hardware
- [x] Add automated tests cases
- [x] Verify that all (other) automated tests (still) pass
- [x] Update the README and other applicable documentation pages
- [ ] Update the `liquidctl.8` Linux/Unix/Mac OS man page
- [x] Update or add applicable `docs/*guide.md` device guides
- [x] Submit relevant data, scripts or dissectors to https://github.com/liquidctl/collected-device-data

New CLI flag?

- [ ] Adjust the completion scripts in `extra/completions/`

New device?

- [ ] Regenerate `extra/linux/71-liquidctl.rules` (instructions in the file header)
- [ ] Add entry to the README's supported device list with applicable notes (at least `en`)

New driver?

- [ ] Document the protocol in `docs/developer/protocol/`

[development process]: https://github.com/liquidctl/liquidctl/blob/main/docs/developer/process.md
[style guide]: https://github.com/liquidctl/liquidctl/blob/main/docs/developer/style-guide.md

### CODE REVIEW

**Comment by jonasmalacofilho** [comment]
_(just copying what we already talked about in the Discord server, so Marshall can find everything in a single place)_

One suggestion I have is to maybe use a context manager (pseudo code):

```py
def get_status(self, **kwargs):
    with self._hardware_fallback():
        for i, speed in enumerate(self._get_speeds()):
            # ...
        # ...
        return status

@contextmanager
def _hardware_fallback(self):
    try:
        self._send_command(_CMD_WAKE)
        yield self
    finally:
        self._send_command(_CMD_SLEEP)
```

Well, the example above is not great: it's weird that something called `_hardware_fallback()` sends the `_CMD_WAKE`...

The other thing is that it may be sensible to add a couple of tests to ensure that both `initialize()` and `get_status()` will indeed fallback to hardware mode (and to make sure we don't break this in the future).

File: liquidctl/driver/commander_core.py [hunk 1]
```diff
@@ -9,10 +9,11 @@
 SPDX-License-Identifier: GPL-3.0-or-later
 """
 
+from contextlib import contextmanager
```
Comment (jonasmalacofilho) [inline liquidctl/driver/commander_core.py:12]: Nit: `import ...` before `from ... import ...`.

File: liquidctl/driver/commander_core.py [hunk 2]
```diff
@@ -140,6 +138,17 @@ def _get_temps(self):
 
         return temps
 
+    def _read_data(self, mode, data_type):
+        self._send_command(_CMD_RESET)
+        self._send_command(_CMD_SET_MODE, mode)
+        res = self._send_command(_CMD_GET)
+
+        input_type = tuple(x for x in res[3:5])
+        if input_type != data_type:
```
Comment (jonasmalacofilho) [inline liquidctl/driver/commander_core.py:147]:
```suggestion
        if tuple(res[3:5]) != data_type:
```

File: liquidctl/driver/commander_core.py [hunk 3]
```diff
@@ -140,6 +138,17 @@ def _get_temps(self):
...
+            raise ExpectationNotMet('Device returned incorrect data type')
```
Comment (jonasmalacofilho) [inline liquidctl/driver/commander_core.py:148]: Nit: error messages should start with a lowercase letter.

File: liquidctl/driver/commander_core.py [hunk 4]
```diff
@@ -43,50 +49,46 @@ class CommanderCore(UsbHidDriver):
...
-            num_leds = u16le_from(res, offset=8+i*4)
-            label = 'AIO LED count' if i == 0 else f'RGB port {i} LED count'
-            status += [(label, num_leds if connected else None, '')]
-
-        # Get what temp sensors are connected
-        for i, temp in enumerate(self._get_temps()):
-            connected = temp is not None
-            label = 'Water temperature sensor' if i == 0 else f'Temperature sensor {i}'
-            status += [(label, connected, '')]
+            num_devices = raw_data[0]
```
Comment (jonasmalacofilho) [inline liquidctl/driver/commander_core.py:62]: I find `raw_data` slightly confusing for someone trying to quickly match the code to a capture, given that it isn't the raw response (which is what I would expect based on the name).

That said, I think you're right to strip out the boilerplate in `_read_data()`, so maybe just invert the names: use `res` in the callers (e.g. here) and `raw_data` inside `_read_data()`.

**Review by jonasmalacofilho — COMMENTED** [review]
Thanks!  Looks good to me, and I only spotted a few minor code style issues.

P.S. I had previously misunderstood the purpose of the `_CMD_SLEEP` (but I don't think it changed anything).

**Comment by ParkerMc** [comment]
> P.S. I had previously misunderstood the purpose of the `_CMD_SLEEP` (but I don't think it changed anything).

Would it be better to name it something else? Wake and sleep are just the two best names I could come up with.

**Comment by jonasmalacofilho** [comment]
> Would it be better to name it something else? Wake and sleep are just the two best names I could come up with.

No need, I think they are fine.

**Review by jonasmalacofilho — APPROVED** [review]

**Comment by jonasmalacofilho** [comment]
Thanks again!

**Comment by ParkerMc** [comment]
Thank you for all the guidance and the base to build off of!

**Pull request merged**